src/httpserver.rs

```diff
@@ -1,178 +1,228 @@
+    if !valid_github_reponame(repo) {
+        return Err(format!("Invalid GitHub repository '{}'.", &repo).into());
     }
 
     let conf = snare.conf.lock().unwrap();
-    let (rconf, secret) = conf.github.repoconfig(&owner, &repo);
+    let (rconf, secret) = conf.github.repoconfig(owner, repo);
 
     match (secret, sig) {
         (Some(secret), Some(sig)) => {
-            if !authenticate(secret, sig, pl) {
-                snare.error(&format!("Authentication failed for {}/{}.", owner, repo));
-                *res.status_mut() = StatusCode::UNAUTHORIZED;
-                return Ok(res);
+            if !authenticate(secret, sig, &body) {
+                return Err(format!("Authentication failed for {}/{}.", owner, repo).into());
             }
         }
         (Some(_), None) => {
-            snare.error(&format!("Request was unsigned for {}/{}.", owner, repo));
-            *res.status_mut() = StatusCode::UNAUTHORIZED;
-            return Ok(res);
+            return Err("Secret specified but request unsigned".into());
         }
         (None, Some(_)) => {
-            snare.error(&format!(
+            return Err(format!(
                 "Request was signed but no secret was specified for {}/{}.",
                 owner, repo
-            ));
-            *res.status_mut() = StatusCode::UNAUTHORIZED;
-            return Ok(res);
+            )
+            .into());
         }
         (None, None) => (),
     }
+    drop(conf);
 
     if event_type == "ping" {
-        *res.status_mut() = StatusCode::OK;
-        return Ok(res);
+        return Ok(());
     }
 
     let repo_id = format!("github/{}/{}", owner, repo);
-    let qj = QueueJob::new(repo_id, owner, repo, req_time, event_type, json_str, rconf);
+    let qj = QueueJob::new(
+        repo_id,
+        owner.to_owned(),
+        repo.to_owned(),
+        req_time,
+        event_type.to_owned(),
+        json_str,
+        rconf,
+    );
     (*snare.queue.lock().unwrap()).push_back(qj);
-    *res.status_mut() = StatusCode::OK;
     // If the write fails, it almost certainly means that the pipe is full i.e. the runner
     // thread will be notified anyway. If something else happens to have gone wrong, then
     // we (and the OS) are probably in deep trouble anyway...
     nix::unistd::write(snare.event_write_fd, &[0]).ok();
-    Ok(res)
+    Ok(())
 }
 
-/// Extract the string 'def' from "X-Hub-Signature: abc=def"
-fn get_hub_sig(req: &Request<Body>) -> Option<String> {
-    req.headers()
-        .get("X-Hub-Signature")
-        .and_then(|s| match s.to_str() {
-            Ok(s) => Some(s),
-            Err(_) => None,
-        })
-        .and_then(|s| s.split('=').nth(1))
-        .map(|s| s.to_owned())
+/// A very literal, and rather unforgiving, implementation of RFC2616 (HTTP/1.1), returning the URL
+/// of GET requests: returns `Err` for anything else.
+fn parse_get(stream: &mut TcpStream) -> Result<(HashMap<String, String>, Vec<u8>), Box<dyn Error>> {
```

Comment:
This function is complicated and is bound to have bugs. It could benefit from unit testing.

Perhaps `TcpStream` is `Read` and this could be generalised to accepting `&str`s and `TcpStream`s, allowing easier testing?

Reply:
It can't accept `&str`s because it has to do some interpretation (notably of `Content-Length`) before carrying on reading. [My first version didn't do that, but github's HTTP request library doesn't do "the right thing" if you try and simplify this!]

The good news is that the new system tests do partially test this. Not completely, of course, and I want to add more tests, but it is at least somewhat tested.

Reply:
I suppose in a way, by connecting to localhost and writing a string down a socket, you are somewhat accepting a string anyway.

I won't push, but it'd be good to test this function in isolation, if not now, later.

Reply:
It's worth stating that the only way to test this function is by passing it data over a socket. The way it has to work is first read the headers and no more. Interpret the headers. If there's `Content-Length`, read *exactly* that number of bytes, no more, no less (this is not theoretical: if you don't do this, GitHub's requests won't work!). So it has to read from a stream, not a string.

Reply:
sure, but you could split the tcp reading from the logic. The logic could accept a tuple of `headers: &str, data: &str`, then you can test the inner.

Reply:
It's worse than that: you *have* to parse the headers before you try reading the body from the network. [This was news to me, FWIW, and cost me a couple of hours debugging yesterday!]